### DISCUSSION TITLE
Why does Pyodide download package whl's every time I refresh a page.

### DISCUSSION THREAD

**Discussion by anchit-sadana** [pyodide/pyodide#2798] [Q&A]
We are currently using pyodide at [Hal9](www.hal9.com). While I was looking at the network usage in developer tools in chrome, it seems that all python whls seem to be downloaded each time I open a page that uses those packages rather than building them from memory or disk cache. Does anyone know why this might be the case? How can I cache the whls and tell pyodide to build the packages from the cache instead of downloading ~5-6 Mb whls every time I open a page that has numpy/pandas imported?
<img width="1552" alt="Screen Shot 2022-06-27 at 3 15 57 PM" src="https://user-images.githubusercontent.com/70300199/176044846-b7f2fc3f-7dec-428e-bdd4-1d686b668e4e.png">

**Comment by rth** [comment]
Can you please click on one of those files and indicate what is the URL they are downloaded from? When downloaded from the JsDelivr CDN they should be cached by the browser and this looks like it's not the case.

**Reply by anchit-sadana** [reply]
Looks like they were delivered by the JsDelivr CDN. For example this the url for the `numpy` whl - `https://cdn.jsdelivr.net/pyodide/v0.20.0/full/numpy-1.22.3-cp310-cp310-emscripten_wasm32.whl`

**Reply by anchit-sadana** [reply]
@rth Is this possibly related https://github.com/pyodide/pyodide/issues/2662? Sorry I am super new to working with web technologies so I am not quite sure.

**Reply by ryanking13** [reply]
#2662 is about caching compiled WASM modules, so I don't think it is related.

**Reply by rth** [reply]
Could you click on one of those files in the F12 download console, and check the response headers? They should look like something,
```
HTTP/3 200 OK
[...]
cache-control: public, max-age=31536000
[...]
last-modified: Sat, 09 Apr 2022 22:11:04 GMT
date: Wed, 29 Jun 2022 07:53:02 GMT
age: 117558
x-served-by: cache-fra19126-FRA, cache-cdg20774-CDG
x-cache: MISS, HIT
[...]
```
Particularly `cache-control` indicates the content is catchable. `last-modified` and `age` indicates that the version cached by the browser was used. While all the `x-*` keys are likely set by the CDN, so not directly related but still interested to know (as maybe one of the CDN providers used by JSDelivr doesn't  pass headers correctly, though it's unlikely)

**Reply by anchit-sadana** [reply]
The response was header was
```
accept-ranges: bytes
access-control-allow-origin: *
access-control-expose-headers: *
age: 1624801
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
cache-control: public, max-age=31536000
cf-cache-status: HIT
cf-ray: 722f654f4a4a844d-YVR
content-length: 5210117
content-type: binary/octet-stream
cross-origin-resource-policy: cross-origin
date: Wed, 29 Jun 2022 14:36:40 GMT
etag: "6f679d66254b93db471d573e7b2d4ac4"
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
last-modified: Sat, 09 Apr 2022 22:11:05 GMT
nel: {"success_fraction":0.01,"report_to":"cf-nel","max_age":604800}
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=NMuzFUvIURgtrYw9aiFzMGQJY%2FptthbxRDDHONaUPja2IRmdxVV5G9l%2B694uek62aaZfGH01wMH38WPqzy9bqObFSJq7rUMC48SCBtEU6hji%2BgzoBfEXf6hWDDWM4njl%2B3b93KCAaLXvUxWdEfo%3D"}],"group":"cf-nel","max_age":604800}
server: cloudflare
strict-transport-security: max-age=31536000; includeSubDomains; preload
timing-allow-origin: *
vary: Accept-Encoding
x-cache: HIT, HIT
x-content-type-options: nosniff
x-served-by: cache-fra19162-FRA, cache-itm18823-ITM
```
